services/ui_backend_service/data/cache/client/cache_server.py

```diff
@@ -257,16 +265,64 @@ def queued_request(queue):
             send_message(OP_WORKER_TERMINATE, worker._worker_details())
             return None
 
+    def verify_stale_workers(self):
+        time_to_pool_refresh = CACHE_PROCESS_POOL_REFRESH_TIME - (time.time() - self._pool_started_on)
+        # active_pids = ",".join([f"{c.name}[{c.pid}]" for c in multiprocessing.active_children()])
+        echo(
+            "number of workers: %d, number of pending requests: %d; Pool Refresh in : %d" % (
+                len(self.workers), len(self.pending_requests), time_to_pool_refresh)
+        )
+
+    def cleanup_if_necessary(self):
+        time_to_pool_refresh = CACHE_PROCESS_POOL_REFRESH_TIME - (time.time() - self._pool_started_on)
+        if time_to_pool_refresh > 0:
+            return
+        # if workers are still running 30 seconds after the pool refresh timeout, then cleanup
+        no_workers_are_running = len(self.workers) == 0 and len(self.pending_requests) == 1
+        pool_needs_refresh = time_to_pool_refresh <= 0
+        pool_force_refresh = time_to_pool_refresh < -60
+        if pool_force_refresh:
+            echo("Refreshing the pool as no workers are running and no pending requests are there.")
+            self.cleanup_workers()
+        elif no_workers_are_running and pool_needs_refresh:
+            echo("Refreshing the pool as no workers are running and no pending requests are there.")
```

Comment:
is there no race here? Can a worker not "get busy" in the meantime?

Reply:
Given that this entire execution is running _sync_ code, and this code block comes after the code path that creates new workers, the only current case is that workers are already created. No _new_  worker will be created during the execution of this block. If workers exist, they are already busy workers. 

Generally, if workers are busy at the time of pool-refresh, I didn't want to kill them immediately. If the server is under a lot of load, then an API request on the server may be waiting for some object from the server whose worker may get killed, and will result in the API response being an ugly http-500x.  

I have not exactly explained in this PR description how the memory leak came to be so let me more light: 
1. When API requests come to the server, the UI Server request the cache-server (running in the background) to extract out the cached value. This request is given a stream-id. The API-server is waiting on the response of the stream-id from the cache-server.
2. The cache server spawns a new worker for this stream-id and worker does some computing and returns the object in the stream
3. When many requests hit the server (especially when a run has a lot of logs), the cache server will spawn multiple individual workers, The API server will be waiting on all of these streams. If the objects(ex. logs) are large, then it will end up loading the whole object in memory and give the response. After the response is passed, the callback handles the removal of the worker and committing information to the cache-store. Even if the worker has been liked, the pool is still holding the reference to it (since we never close/join the pool)